### PR TITLE
make fileStore a protected field for Job (resolves #1224)

### LINKED ISSUE #1224: All jobs should have access to the underlying fileStore
From #1198, [comment](https://github.com/BD2KGenomics/toil/pull/1198#discussion_r84343466). Right now, only `ServiceJob` and `JobFunctionWrappingJob` set `self.fileStore`. We should make this consistent across jobs.

### PR DESCRIPTION
**Pull request by jessebrennan**
The purpose of this PR is to give all jobs access to the underlying fileStore but as a protected class attribute. Before, serviceJob and jobFunctionWrapingJob needed access to the fileStore and the thought was to give access to all jobs as a class attribute. But this creates two sorts of truths—the jobStore that a user passes in to run, and the one associated with the class. By making it protected, it can be safely ignored and is available if the user knows what they are doing. ServiceJob and jobFunctionWrappingJob have a property to mask the protected class to keep one source of truth since they need explicit access to the class attribute.

Changes:
- The protected attribute `_fileStore` is added to `Job` and is set in Job._runner
- `Job.Defer` now uses correct fileStore attribute.
- a `fileStore` property is used in `JobFunctionWrappingJob` and `ServiceJob` to mask `Job._fileStore` to keep one source of truth

### CODE REVIEW

File: src/toil/job.py [hunk 1]
```diff
@@ -1381,6 +1378,10 @@ class JobFunctionWrappingJob(FunctionWrappingJob):
     instance (see :func:`toil.job.Job.Run`), it is made a variable of the wrapping job \
     called fileStore.
     """
+    @property
+    def fileStore(self):
+        return self._fileStore
+
     def run(self, fileStore):
         userFunction = self._getUserFunction()
         self.fileStore = fileStore
```
Comment (hannes-ucsc) [inline src/toil/job.py]: This is obviously redundant. Same in ServiceJob.

File: src/toil/job.py [hunk 2]
```diff
@@ -1271,7 +1268,7 @@ def _runner(self, jobGraph, jobStore, fileStore):
         :return:
         """
         # Run the job
-        returnValues = self._run(jobGraph, fileStore)
+        returnValues = self.run(fileStore)
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Shouldn't this also set `self._fileStore`?

File: src/toil/job.py (hunk 1 above)
Comment (jessebrennan) [inline src/toil/job.py]: I had thought this was the agreed solution. Are you suggesting that I add the public fileStore property to the Job class?

**Comment by jessebrennan** [comment]
@hannes-ucsc please review

**Comment by hannes-ucsc** [comment]
I'd prefer to review after branch is rebased and conflicts have been resolved.

File: src/toil/job.py [hunk 3]
```diff
@@ -289,6 +289,8 @@ def __init__(self, memory=None, cores=None, disk=None, preemptable=None, unitNam
         # entire return value.
         self._rvs = collections.defaultdict(list)
         self._promiseJobStore = None
+        # All jobs have access to protected fileStore field. Especially needed for defer
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Remove comment. First sentence is obvious. Second sentence is anecdotal.

File: src/toil/job.py [hunk 4]
```diff
@@ -1270,6 +1270,8 @@ def _runner(self, jobGraph, jobStore, fileStore):
                filestore
         :return:
         """
+        # store fileStore for use in JobFunctionWrappingJob, ServiceJob, and defer
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Change to `# Make fileStore available as an attribute during run() ...`

File: src/toil/job.py [hunk 5]
```diff
@@ -1270,6 +1270,8 @@ def _runner(self, jobGraph, jobStore, fileStore):
...
+        self._fileStore = fileStore
         # Run the job
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Replace with `# ... but also pass it to run() as an argument for backwards compatibility.`

**Comment by jessebrennan** [comment]
@cket I think this is ready to go. Please review.

**Comment by benedictpaten** [comment]
Can we get this in @hannes-ucsc @cket

**Comment by cket** [comment]
Jenkins, test this please

**Comment by jessebrennan** [comment]
@cket ready for review.

**Comment by cket** [comment]
@jessebrennan thanks!

**Comment by jessebrennan** [comment]
SO this "fix" is completely wrong (as I understand it now). It was based on the assumption that there would be multiple sources of "truth" for the actual filestore. This is *not* the case. I am going to reopen the original issue.

**Pull request merged**